src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,164 @@
+import logging
+import re
+from typing import Dict, List, Set, Union, cast
+
+from ruamel.yaml.comments import CommentedMap, CommentedSeq
+from ruamel.yaml.emitter import Emitter
+
+# Module 'ruamel.yaml' does not explicitly export attribute 'YAML'; implicit reexport disabled
+# To make the type checkers happy, we import from ruamel.yaml.main instead.
+from ruamel.yaml.main import YAML
+
+from .errors import MatchError
+from .file_utils import Lintable
+from .runner import LintResult
+from .skip_utils import load_data  # TODO: move load_data out of skip_utils
+
+__all__ = ["Transformer"]
+
+_logger = logging.getLogger(__name__)
+
+# ruamel.yaml only preserves empty (no whitespace) blank lines
+# (ie "/n/n" becomes "/n/n" but "/n  /n" becomes "/n").
+# So, we need to identify whitespace-only lines to drop spaces before reading.
+_whitespace_only_lines_re = re.compile(r"^ +$", re.MULTILINE)
+
+
+class FormattedEmitter(Emitter):
+    """Emitter that applies custom formatting rules when dumping YAML.
+
+    Root-level sequences are never indented.
+    All subsequent levels are indented as configured (normal ruamel.yaml behavior).
+
+    Earlier implementations used dedent on ruamel.yaml's dumped output,
+    but string magic like that had a ton of problematic edge cases.
+    """
+
+    _sequence_indent = 2
+    _sequence_dash_offset = 0  # Should be _sequence_indent - 2
+
+    # NB: mypy does not support overriding attributes with properties yet:
+    #     https://github.com/python/mypy/issues/4125
+    #     To silence we have to ignore[override] both the @property and the method.
+
+    @property  # type: ignore[override]
+    def best_sequence_indent(self) -> int:  # type: ignore[override]
+        """Return the configured sequence_indent or 2 for root level."""
+        return 2 if self.column < 2 else self._sequence_indent
+
+    @best_sequence_indent.setter
+    def best_sequence_indent(self, value: int) -> None:
+        """Configure how many columns to indent each sequence item (including the '-')."""
+        self._sequence_indent = value
+
+    @property  # type: ignore[override]
+    def sequence_dash_offset(self) -> int:  # type: ignore[override]
+        """Return the configured sequence_dash_offset or 2 for root level."""
+        return 0 if self.column < 2 else self._sequence_dash_offset
+
+    @sequence_dash_offset.setter
+    def sequence_dash_offset(self, value: int) -> None:
+        """Configure how many spaces to put before each sequence item's '-'."""
+        self._sequence_dash_offset = value
+
+
+# Transformer is for transforms like runner is for rules
+class Transformer:
+    """Transformer class performs the fmt transformations."""
+
+    def __init__(self, result: LintResult):
+        """Initialize a Transformer instance."""
+        # TODO: options for explict_start, indent_sequences
+        self.explicit_start = True
+        self.indent_sequences = True
+
+        self.matches: List[MatchError] = result.matches
```

Comment:
It looks like this to the end of the function is calculating the matches per file. Should computing that be the responsibility of the Transformer?

Reply:
I would disregard this comment now. It makes more sense for the "singleton Transformer that runs all the TransformerMixin.transform" to keep track of this.